<div dir="rtl">
    
# مثال على استخدام hkvwaporpy
    
</div>

<div dir="rtl">
في هذا الدفتر، سنستخدم حزمة hkvwaporpy لجمع بيانات WaPOR في Python. أولاً ، استيراد الحزم الإضافية: *hkvwaporpy*, *requests*, *os*, و *datetime*

</div>

In [1]:
import hkvwaporpy as hkv
import requests
import datetime
import os

<div dir="rtl">
    
## 1. كتالوج قاعدة البيانات/cubes
    
</div>

<div dir="rtl">
تسمح الحزمة hkvwaporpy بجلب قائمة مجموعات البيانات في قاعدة بيانات WaPOR باستخدام الوظيفة get_catalogus(). تحتوي كل مجموعة بيانات أو مكعب بيانات على رمز مكعب محدد. على سبيل المثال، رمز المكعب للمستوى 1 هطول الأمطار (يوميًا) هو L1_PCP_E. قم بتشغيل خلية التعليمات البرمجية أدناه، ثم حاول تغيير قيمة الإصدار المعلمة ل "2.0" لتحديد WaPOR الإصدار 2.0.
</div>

In [2]:
df=hkv.read_wapor.get_catalogus(version='1.1') 
df[['caption','code']]

caption                code
0                     Gross Biomass Water Productivity           L1_GBWP_A
1                       Net Biomass Water Productivity           L1_NBWP_A
2    Actual EvapoTranspiration and Interception (An...           L1_AETI_A
3    Actual EvapoTranspiration and Interception (De...           L1_AETI_D
4                               Transpiration (Annual)              L1_T_A
5                                 Evaporation (Annual)              L1_E_A
6                                Interception (Annual)              L1_I_A
7                              Transpiration (Dekadal)              L1_T_D
8                                Evaporation (Dekadal)              L1_E_D
9                               Interception (Dekadal)              L1_I_D
10                              Net Primary Production            L1_NPP_D
11                     Above Ground Biomass Production           L1_AGBP_A
12                           Land Cover Classification            L1_LCC_A
13               Reference EvapoTranspiration (Annual)            L1_RET_A
14                              Precipitation (Annual)            L1_PCP_A
15              Reference EvapoTranspiration (Dekadal)            L1_RET_D
16                             Precipitation (Dekadal)            L1_PCP_D
17                Reference EvapoTranspiration (Daily)            L1_RET_E
18                               Precipitation (Daily)            L1_PCP_E
19      Quality Normalized Difference Vegetation Index      L1_QUAL_NDVI_D
20                    Quality Land Surface Temperature       L1_QUAL_LST_D
21         Gross Biomass Water Productivity (Seasonal)           L2_GBWP_S
22           Net Biomass Water Productivity (Seasonal)           L2_NBWP_S
23   Actual EvapoTranspiration and Interception (An...           L2_AETI_A
24   Actual EvapoTranspiration and Interception (De...           L2_AETI_D
25                              Transpiration (Annual)              L2_T_A
26                                Evaporation (Annual)              L2_E_A
27                               Interception (Annual)              L2_I_A
28                             Transpiration (Dekadal)              L2_T_D
29                               Evaporation (Dekadal)              L2_E_D
..                                                 ...                 ...
114  Net Biomass Water Productivity (Seasonal, clip...       L2_NBWP_BAS_S
115  Net Biomass Water Productivity (Seasonal, clip...       L2_NBWP_CTY_S
116  Actual EvapoTranspiration and Interception (An...       L2_AETI_BAS_A
117  Actual EvapoTranspiration and Interception (An...       L2_AETI_CTY_A
118  Actual EvapoTranspiration and Interception (De...       L2_AETI_BAS_D
119  Actual EvapoTranspiration and Interception (De...       L2_AETI_CTY_D
120          Transpiration (Annual, clipped by basins)          L2_T_BAS_A
121         Transpiration (Annual, clipped by country)          L2_T_CTY_A
122            Evaporation (Annual, clipped by basins)          L2_E_BAS_A
123           Evaporation (Annual, clipped by country)          L2_E_CTY_A
124           Interception (Annual, clipped by basins)          L2_I_BAS_A
125          Interception (Annual, clipped by country)          L2_I_CTY_A
126         Transpiration (Dekadal, clipped by basins)          L2_T_BAS_D
127        Transpiration (Dekadal, clipped by country)          L2_T_CTY_D
128           Evaporation (Dekadal, clipped by basins)          L2_E_BAS_D
129          Evaporation (Dekadal, clipped by country)          L2_E_CTY_D
130          Interception (Dekadal, clipped by basins)          L2_I_BAS_D
131         Interception (Dekadal, clipped by country)          L2_I_CTY_D
132  Net Primary Production (Dekadal, clipped by ba...        L2_NPP_BAS_D
133  Net Primary Production (Dekadal, clipped by co...        L2_NPP_CTY_D
134  Above Ground Biomass Production (Seasonal, cli...       L2_AGBP_BAS_S
135  Above Ground Biomass Production (Seasonal, cli...       L2_AGBP_CTY

<div dir="rtl">
    
## 2.معلومات مجموعة البيانات / المكعب
    
</div>


<div dir="rtl">    
تحصل الحزمة hkvwaporpy أيضًا على معلومات مجموعة البيانات باستخدام الوظيفة
get_info_cube برمز مكعب معين. تسمى مجموعة البيانات بالمكعب لأنها تحتوي على 3 أبعاد (خط العرض وخط الطول والوقت). على سبيل المثال ، الكود المكعب للمستوى 1 Dekadal Reference EvapoTranspiration هو L1_RET_D بالنسبة لمجموعة البيانات الأخرى ، تحقق من قائمة كتالوج الرموز.    
</div>

In [3]:
ds_code='L1_RET_D' ## كود المكعب للمستوى 1 Dekadal Reference البخرنتح
cube_info=hkv.read_wapor.get_info_cube(cube_code=ds_code)
print(cube_info)

                                                                 L1_RET_D
format                                                     Raster Dataset
unit                                                                   mm
dataType                                                             Byte
conversionFactor        the pixel value in the downloaded data must be...
noDataValue                                                           255
spatialResolution                        Approximately 20km (0.17 degree)
spatialExtent                                        Africa and Near East
spatialReferenceSystem  EPSG:4326 - WGS84 - Geographic Coordinate Syst...
temporalResolution                           from January 2009 to present
temporalExtent                      Dekadal (approximately every 10 days)
methodology             See daily Reference EvapoTranspiration (L1_RET...
dimensions              code                                          ...
measures                code          

<div dir="rtl">
عند طباعة معلومات cube_info ، تظهر تفاصيل مجموعة البيانات، مثل التنسيق ونوع البيانات والمدى المكاني والدقة وما إلى ذلك. لمشاهدة المزيد من التفاصيل من cube_info، استخدم cube_info.loc['detail',ds_code]. على سبيل المثال، تعطي الأبعاد معلومات حول الدقة الزمنية (الخطوة الزمنية) وتوفر المقاييس معلومات حول الوحدة والقيمة المتعددة لتحويل الرقم الرقمي (البيانات الأولية) إلى القيمة المادية.
</div>

In [4]:
cube_info.loc['dimensions',ds_code]

code                                                       DEKAD
caption                                   Dekad (10-Days period)
code                                                       DEKAD
description    Each month is splitted in 3 10-Days periods: f...
hierarchical                                               False
index                                                          1
links          [{'rel': 'self', 'href': 'https://io.apps.fao....
timeSubtype                                              10-DAYS
type                                                        TIME
workspaceCode                                              WAPOR
season                                                       NaN
stage                                                        NaN

In [5]:
cube_info.loc['measures',ds_code]

code                                                    WATER_MM
caption                                          Amount of Water
code                                                    WATER_MM
description    Amount of water expressed in mm which can be c...
index                                                          2
links          [{'rel': 'self', 'href': 'https://io.apps.fao....
multiplier                                                   0.1
scale                                                          3
unit                                                          mm
workspaceCode                                              WAPOR

<div dir="rtl">
    
## 3. توافر البيانات ومعرف البيانات النقطية

</div>

<div dir="rtl">
للحصول على قائمة بالبيانات النقطية المتوفرة في مجموعة البيانات / المكعب لفترة، استخدم الدالة get_data_availability مع معلومات cube_info التي تم الحصول عليها. سيتم تحديد كل خريطة نقطية بواسطة معرف البيانات النقطية (raster_id). على سبيل المثال، يتم استخدام الكود أدناه للحصول على المستوى 1 Dekadal الفعلي لبيانات البخرنتح والاعتراض من 2008-01-01 إلى 2018-12-31. هنا ، لاحظ أن بيانات WaPOR متاحة فقط من عام 2009.
</div>

In [6]:
df_avail=hkv.read_wapor.get_data_availability(cube_info=cube_info,time_range='[2008-01-01,2018-12-31]')
df_avail[['raster_id']]

data_avail_period: DEKAD


raster_id
year             
2009  L1_RET_0901
2009  L1_RET_0902
2009  L1_RET_0903
2009  L1_RET_0904
2009  L1_RET_0905
2009  L1_RET_0906
2009  L1_RET_0907
2009  L1_RET_0908
2009  L1_RET_0909
2009  L1_RET_0910
2009  L1_RET_0911
2009  L1_RET_0912
2009  L1_RET_0913
2009  L1_RET_0914
2009  L1_RET_0915
2009  L1_RET_0916
2009  L1_RET_0917
2009  L1_RET_0918
2009  L1_RET_0919
2009  L1_RET_0920
2009  L1_RET_0921
2009  L1_RET_0922
2009  L1_RET_0923
2009  L1_RET_0924
2009  L1_RET_0925
2009  L1_RET_0926
2009  L1_RET_0927
2009  L1_RET_0928
2009  L1_RET_0929
2009  L1_RET_0930
...           ...
2018  L1_RET_1807
2018  L1_RET_1808
2018  L1_RET_1809
2018  L1_RET_1810
2018  L1_RET_1811
2018  L1_RET_1812
2018  L1_RET_1813
2018  L1_RET_1814
2018  L1_RET_1815
2018  L1_RET_1816
2018  L1_RET_1817
2018  L1_RET_1818
2018  L1_RET_1819
2018  L1_RET_1820
2018  L1_RET_1821
2018  L1_RET_1822
2018  L1_RET_1823
2018  L1_RET_1824
2018  L1_RET_1825
2018  L1_RET_1826
2018  L1_RET_1827
2018  L1_RET_1828
2018  L1_RET_1829
2018  L1_RET_1830
2018  L1_RET_1831
2018  L1_RET_1832
2018  L1_RET_1833
2018  L1_RET_1834
2018  L1_RET_1835
2018  L1_RET_1836

[360 rows x 1 columns]

<div dir="rtl">
في قائمة data_avail المتاحة، يمكن الوصول إلى raster_id من رقم الفهرس. على سبيل المثال، النقطية الأولى مع الفهرس i = 0 لها raster_id
L1_RET_0901. يتم استخدام هذا المعرف لاحقًا لطلب عنوان URL للتنزيل من قاعدة البيانات.
</div>

In [7]:
i=0 #احصل على أول طبقة راستر من df_avail
cube_period=df_avail.iloc[i]
raster_id=cube_period['raster_id']
print(raster_id)

L1_RET_0901


<div dir="rtl">
    
## 4. احصل على رابط التحميل

</div>
<div dir="rtl">     
للحصول على رابط تنزيل البيانات النقطية (أو مجموعة الخرائط) ، استخدم الوظيفة get_coverage_url. تتطلب هذه الوظيفة كود adataset / cube ، ومعرف نقطي ، ورمز API صالحًا ، يتم الحصول عليه من
(https://wapor.apps.fao.org/profile)
باستخدام حساب WaPOR (راجع دفتر الملاحظات 0_Start_here (0_Start_here.ipynb) للحصول على إرشادات حول كيفية الحصول على رمز API المميز)
</div>
<div dir="rtl">
قم بتعيين قيمة متغير api_token في خلية الكود أدناه عن طريق استبدال Your_WaPOR_API_token برمز yourapi الذي تم الحصول عليه من ملف تعريف WaPOR. تذكر أن تضعه بين علامتي اقتباس مفردتين لتعريفهما كمتغير سلسلة.
</div>

In [8]:
# احصل على رمز الAPI الخاص بك https://wapor.apps.fao.org/profile
api_token='Your_WaPOR_API_token'

<div dir="rtl">    
من النتائج السابقة ، ds_code هنا هو L1_RET_D و raster_id هنا هو L1_RET_0901. ستعيد الوظيفة عنوان url لتنزيل download_url مع انتهاء وقت الصلاحية expiry_datetime. إذا كان رمز API المقدم في الخلية السابقة صحيحًا ، فسيكون hkvwaporpy قادرًا على جمع الارتباط (url) لتنزيل البيانات النقطية. إذا لم يكن الأمر كذلك ، فستُرجع خلية الرمز أدناه "استجابة" KeyError ، مما يعني عدم وجود استجابة من الخادم.
</div>

In [9]:
cov_object = hkv.read_wapor.get_coverage_url(
    APItoken = api_token, 
    raster_id = raster_id,
    cube_code = ds_code)
print(cov_object)

{'expiry_datetime': datetime.datetime(2020, 9, 21, 11, 46, 30, 190364), 'download_url': 'https://io.apps.fao.org/gismgr/download/1fd5bac3-be18-4789-8909-798a5465cd0e/L1_RET_0901.tif'}


<div dir="rtl">

## 5. قم بتنزيل مجموعة الخرائط النقطية من المستوى 1
    
</div>
<div dir="rtl">
يمكن استخدام عنوان url الذي تم الحصول عليه من وظيفة get_coverage_url لتنزيل الخريطة النقطية مباشرةً من قاعدة البيانات باستخدام request.get. على سبيل المثال ، يتم استخدام الكود أدناه لتنزيل الخريطة النقطية L1_RET_0901.tif إلى المجلد r'.\data\L1_RET_D'.    
</div>

[**r'.\data\L1_RET_D'**](data)    

<div dir="rtl">    
    لتحديد مجلد مختلف لحفظ البيانات ، قم بتغيير قيمة المجلد output_folder.
</div>

In [10]:
output_folder=r'.\data\L1_RET_D' #ملف لحفظ الداتا
if not os.path.exists(output_folder): #تأكد أن كان output_folder صحيحاً
    os.makedirs(output_folder) #خلق output_folder
filename=output_folder+'\{0}.tif'.format(raster_id) #حفظ ملف مع معرف نقطي كاسم ملف
#تنزيل ملف نقطي
resp=requests.get(cov_object['download_url']) 
open(filename,'wb').write(resp.content) 
print(filename)

.\data\L1_RET_D\L1_RET_0901.tif


<div dir="rtl">
    
## 6. رمز الموقع

</div>
<div dir="rtl">
تتضمن بيانات المستوى 1 من WaPOR خرائط نقطية تغطي القارة الإفريقية بأكملها ومنطقة الشرق الأدنى ، بينما تغطي بيانات المستوى 2 WaPOR بعض البلدان والأحواض فقط. لتنزيل بيانات المستوى 2 ، يجب تحديد رمز الموقع الخاص بالبلد أو الحوض المحدد. يمكن الوصول إلى قائمة رموز الموقع باستخدام وظيفة get_locations (). هنا ، إذا كانت المعلمة L2 للموقع لها قيمة حقيقية ، فهذا يعني أن بيانات المستوى 2 متاحة لهذا الموقع.
</div>

In [11]:
df_locations = hkv.read_wapor.get_locations()
df_locations[['name','type','code','L1','L2','L3']]

name     type  code     L1     L2     L3
0                              Awash    BASIN  7010  False   True  False
1                             Jordan    BASIN  6006  False   True  False
2                             Litani    BASIN  6002  False   True  False
3                              Niger    BASIN  7002  False   True  False
4                               Nile    BASIN  7003  False   True  False
5                            Algeria  COUNTRY   DZA  False  False  False
6                             Angola  COUNTRY   AGO  False  False  False
7                            Bahrain  COUNTRY   BHR  False  False  False
8                              Benin  COUNTRY   BEN  False   True  False
9                           Botswana  COUNTRY   BWA  False  False  False
10                      Burkina Faso  COUNTRY   BFA  False  False  False
11                           Burundi  COUNTRY   BDI  False   True  False
12                        Cabo Verde  COUNTRY   CPV  False  False  False
13                          Cameroon  COUNTRY   CMR  False  False  False
14          Central African Republic  COUNTRY   CAF  False  False  False
15                              Chad  COUNTRY   TCD  False  False  False
16                           Comoros  COUNTRY   COM  False  False  False
17                             Congo  COUNTRY   COG  False  False  False
18                     Côte d'Ivoire  COUNTRY   CIV  False  False  False
19  Democratic Republic of the Congo  COUNTRY   COD  False  False  False
20                          Djibouti  COUNTRY   DJI  False  False  False
21                             Egypt  COUNTRY   EGY  False   True  False
22                 Equatorial Guinea  COUNTRY   GNQ  False  False  False
23                           Eritrea  COUNTRY   ERI  False  False  False
24                          Eswatini  COUNTRY   SWZ  False  False  False
25                          Ethiopia  COUNTRY   ETH  False   True  False
26                             Gabon  COUNTRY   GAB  False  False  False
27                            Gambia  COUNTRY   GMB  False  False  False
28                             Ghana  COUNTRY   GHA  False   True  False
29                            Guinea  COUNTRY   GIN  False  False  False
..                               ...      ...   ...    ...    ...    ...
44                        Mauritania  COUNTRY   MRT  False  False  False
45                         Mauritius  COUNTRY   MUS  False  False  False
46                           Morocco  COUNTRY   MAR  False   True  False
47                        Mozambique  COUNTRY   MOZ  False   True  False
48                           Namibia  COUNTRY   NAM  False  False  False
49                             Niger  COUNTRY   NER  False  False  False
50                           Nigeria  COUNTRY   NGA  False  False  False
51    Occupied Palestinian Territory  COUNTRY   PSE  False   True  False
52                              Oman  COUNTRY   OMN  False  False  False
53                             Qatar  COUNTRY   QAT  False  False  False
54                            Rwanda  COUNTRY   RWA  False   True  False
55             Sao Tome and Principe  COUNTRY   STP  False  False  False
56                      Saudi Arabia  COUNTRY   SAU  False  False  False
57                           Senegal  COUNTRY   SEN  False  False  False
58                        Seychelles  COUNTRY   SYC  False  False  False
59                      Sierra Leone  COUNTRY   SLE  False  False  False
60                           Somalia  COUNTRY   SOM  False  False  False
61                      South Africa  COUNTRY   ZAF  False  False  False
62                       South Sudan  COUNTRY   SSD  False   True  False
63                             Sudan  COUNTRY   SDN  False  False  False
64              Syrian Arab Republic  COUNTRY   SYR  False   True  False
65                              Togo  COUNTRY   TGO  False  False  False
66                           Tunisia  COUNTRY   TUN  False   True  False
67               

<div dir="rtl">
على سبيل المثال، الموقع الذي يحتوي على رقم فهرس 0 هو حوض Awash به بيانات المستوى 2 المتاحة. رمز الموقع هو 7010 ، والذي يمكن استخدامه لطلب مجموعة بيانات المستوى 2 لحوض Awash في الخطوة التالية.
</div>

In [12]:
location=df_locations.iloc[0]
loc_type=location['type']
loc_code=location['code']
print(loc_type,loc_code,location['name'])

BASIN 7010 Awash


<div dir="rtl">

## 7. تنزيل مجموعة البيانات النقطية للموقع من المستوى 2

</div>
<div dir="rtl">
لتنزيل مجموعة البيانات من المستوى 2 ، يجب تحديد رمز مجموعة البيانات / المكعب أولاً ويجب الحصول على معلومات مجموعة البيانات باستخدام الوظيفة
get_info_cube . فمثلا، L2_AETI_D هو رمز المكعب للمستوى 2 DekadalActual البخرنتح و الإعتراض، كما يمكن العثور عليه في الكتالوج من المثال 1.
</div>

In [13]:
ds_code='L2_AETI_D'
cube_info=hkv.read_wapor.get_info_cube(cube_code=ds_code)
print(cube_info)

                                                                L2_AETI_D
format                                                     Raster Dataset
unit                                                                   mm
dataType                                   Int16 (16bit Unsigned Integer)
conversionFactor        the pixel value in the downloaded data must be...
noDataValue                                                         -9999
spatialResolution                                  100m (0.000992 degree)
spatialExtent                                        Africa and Near East
spatialReferenceSystem  EPSG:4326 - WGS84 - Geographic Coordinate Syst...
temporalResolution                           from January 2009 to present
temporalExtent                      Dekadal (approximately every 10 days)
methodology             The calculation of the ETIa is based on the ET...
dimensions              code                                          ...
measures                code          

<div dir="rtl">
باستخدام الوظيفة get_data_availability كما في المثال 3 ، يمكن الحصول على قائمة الخرائط النقطية المتاحة خلال فترة زمنية. على سبيل المثال ، يتم استخدام الكود أدناه للحصول على قائمة معرّف البيانات النقطية لـ L2_AETI_D في يناير 2009. لتحديد فترة زمنية مختلفة ، قم بتغيير قيمة المعلمة time_range.
</div>

In [14]:
df_avail=hkv.read_wapor.get_data_availability(cube_info=cube_info,time_range='[2009-01-01,2009-01-31]')
df_avail[['raster_id']]

data_avail_period: DEKAD


raster_id
year              
2009  L2_AETI_0901
2009  L2_AETI_0902
2009  L2_AETI_0903

<div dir="rtl">
باستخدام الوظيفة get_data_availability كما في المثال 3 ، يمكن الحصول على قائمة الخرائط النقطية المتاحة خلال فترة زمنية. على سبيل المثال ، يتم استخدام الكود أدناه للحصول على قائمة معرّف البيانات النقطية لـ L2_AETI_D في يناير 2009. لتحديد فترة زمنية مختلفة ، قم بتغيير قيمة المعلمة time_range.
</div>

In [15]:
output_folder=r'.\data\L2_AETI_D'
if not os.path.exists(output_folder): #check إن تم عمل ملف للمخرجات
    os.makedirs(output_folder) #قم بانشاء ملف للمخرجات
for i in range(len(df_avail)): # تفقد جميع المعلومات النقطية في قائمة البيانات النقطية المتاحة
    cube_period = df_avail.iloc[i] #إحصل على الطبقة النقطية
    raster_id = cube_period['raster_id'] #هوية الطبقة النقطية
    # اجمع عنوان url لتنزيل البيانات النقطية
    cov_object = hkv.read_wapor.get_coverage_url(
        APItoken=api_token,
        raster_id = raster_id,
        cube_code = ds_code,
        loc_type = loc_type, 
        loc_code = loc_code)
    print(cov_object['download_url']) #اطبع url للتحميل
    if cov_object['expiry_datetime']>datetime.datetime.now(): #تفقد ما إذا كان عنوان url غير منتهي الصلاحية
        # إذا كان عنوان url لا يزال صالحًا ، فقم بتنزيل الملف
        resp = requests.get(cov_object['download_url'])
        open(output_folder+'\{0}.tif'.format(raster_id),'wb').write(resp.content)
        print('Saved to: '+output_folder+'\{0}.tif'.format(raster_id)) #اطبع مسار الملف الذي تم تنزيله

https://io.apps.fao.org/gismgr/download/ff6debdc-a47e-4662-b5b1-3e32922fb317/L2_AETI_0901_7010.tif
Saved to: .\data\L2_AETI_D\L2_AETI_0901.tif
https://io.apps.fao.org/gismgr/download/3ac8411f-82cb-4966-b813-ab3f9a103612/L2_AETI_0902_7010.tif
Saved to: .\data\L2_AETI_D\L2_AETI_0902.tif
https://io.apps.fao.org/gismgr/download/e7ff484d-3c5b-46b5-84bd-23ebdbd54774/L2_AETI_0903_7010.tif
Saved to: .\data\L2_AETI_D\L2_AETI_0903.tif


<div dir="rtl">
    
## التمرين 1
باتباع الخطوات المذكورة أعلاه، هل يمكنك جمع بيانات WaPOR المستوى 1 للهطول اليومي (L1_PCP_E) لعام 2009 وبيانات WaPOR المستوى 2 لحوض Awash لنفس العام للمتغيرات التالية؟

- المستوى 2 Dekadal الفعلي التبخر والاعتراض (L2_AETI_D)
- المستوى 2 تبخر ديكادال (L2_E_D)
- المستوى 2 النتح ديكادال (L2_T_D)
- اعتراض ديكدال المستوى 2 (L2_I_D)    
</div>

In [16]:
"""
اكتب الكود الخاص بك هنا
"""

'\nWrite your code here\n'